apps/emqx_oracle/src/emqx_oracle.erl

```diff
@@ -0,0 +1,453 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_oracle).
+
+-behaviour(emqx_resource).
+
+-include_lib("emqx_resource/include/emqx_resource.hrl").
+-include_lib("emqx_connector/include/emqx_connector.hrl").
+-include_lib("emqx/include/logger.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+-define(ORACLE_DEFAULT_PORT, 1521).
+
+%%====================================================================
+%% Exports
+%%====================================================================
+
+%% callbacks for behaviour emqx_resource
+-export([
+    callback_mode/0,
+    is_buffer_supported/0,
+    on_start/2,
+    on_stop/2,
+    on_query/3,
+    on_batch_query/3,
+    on_query_async/4,
+    on_batch_query_async/4,
+    on_get_status/2
+]).
+
+%% callbacks for ecpool
+-export([connect/1, prepare_sql_to_conn/2]).
+
+%% Internal exports used to execute code with ecpool worker
+-export([
+    query/3,
+    execute_batch/3,
+    do_async_reply/2,
+    do_get_status/1
+]).
+
+-export([
+    oracle_host_options/0
+]).
+
+-import(emqx_plugin_libs_rule, [str/1]).
+-import(hoconsc, [mk/2, enum/1, ref/2]).
+
+-define(ACTION_SEND_MESSAGE, send_message).
+
+-define(SYNC_QUERY_MODE, no_handover).
+-define(ASYNC_QUERY_MODE(REPLY), {handover_async, {?MODULE, do_async_reply, [REPLY]}}).
+
+-define(ORACLE_HOST_OPTIONS, #{
+    default_port => ?ORACLE_DEFAULT_PORT
+}).
+
+-define(MAX_CURSORS, 10).
+-define(DEFAULT_POOL_SIZE, 8).
+-define(OPT_TIMEOUT, 30000).
+
+-type prepares() :: #{atom() => binary()}.
+-type params_tokens() :: #{atom() => list()}.
+
+-type state() ::
+    #{
+        poolname := atom(),
+        prepare_sql := prepares(),
+        params_tokens := params_tokens(),
+        batch_params_tokens := params_tokens()
+    }.
+
+callback_mode() -> async_if_possible.
```

Comment:
we found out that ecpool's async_handover does not work quite well for v5's buffer worker.
we'll have to change to use `always_sync` for now.
```suggestion
callback_mode() -> always_sync.
```